ext/afform/core/Civi/Api4/Action/CustomGroup/GetAfforms.php

```diff
@@ -0,0 +1,375 @@
+<?php
+
+namespace Civi\Api4\Action\CustomGroup;
+
+use CRM_Afform_ExtensionUtil as E;
+use Civi\Api4\Utils\CoreUtil;
+
+/**
+ * Get dynamic afforms for a custom group.
+ *
+ * For single value custom groups we generate:
+ * - a field block with all of the fields from the group
+ * - a submission form for editing these fields on a parent entity record
+ *
+ * For multi-value custom groups we generate:
+ * - a field block with all of the fields from the group
+ * - a submission form for adding a new Custom record to a parent entity record
+ * - a submission form for editing a particular Custom record
+ * - (not implemented yet) a search form for viewing all the Custom records
+ *
+ * @package Civi\Api4\Action\CustomGroup
+ */
+class GetAfforms extends \Civi\Api4\Generic\BasicBatchAction {
+
+  protected const FORM_TYPES = [
+    'afblockCustom',
+    'afformUpdateCustom',
+    'afformCreateCustom',
+    'afsearchTabCustom',
+  ];
+
+  /**
+   * Whether to generate the form layouts
+   * @var bool
+   */
+  protected bool $getLayout = FALSE;
+
+  /**
+   * Limit to generating specific afform types
+   *
+   * Default to all
+   *
+   * @var array
+   */
+  protected array $formTypes = ['block', 'form', 'search'];
+
+  protected function getSelect() {
+    return ['id', 'name', 'title', 'is_multiple',
+      'help_pre', 'help_post', 'extends', 'icon', 'style',
+    ];
+  }
+
+  protected function doTask($item) {
+    $forms = [];
+
+    // get field names once, for use across all the generate actions
+    $item['field_names'] = \Civi\Api4\CustomField::get(FALSE)
+      ->addSelect('name')
+      ->addWhere('custom_group_id', '=', $item['id'])
+      ->addWhere('is_active', '=', TRUE)
+      ->execute()
+      ->column('name');
```

Comment:
Custom fields are cached in memory along with groups. You can use `CRM_Core_BAO_CustomGroup.getGroup()['fields']` to avoid this database hit.